### PR TITLE
cmd/contour: Wait for informers to sync

### LINKED ISSUE #1280: Wait to discover all K8s resources before sending xDS responses
As far as I can tell Contour will send whatever resources it has whenever xDS DiscoveryRequests come in. This can cause xDS resources to be torn down in Envoy on a Contour restart. Contour should do a synchronous (read: eager and blocking) lookup of its CRDs, Ingresses, Services, Endpoints, Secrets, etc. on startup before responding to DiscoveryRequests.

This is causing a variety of issues depending on the resource. Examples include:
1. Missing EDS causes 503s
1. Missing RDS causes 404s
1. Missing SDS causes SSL handshake errors

/cc @bryanlatten @lrouquette

--- 

Tasks

- [x] Wait for k8s informer caches to sync
- [ ] Wait for dag to be fully built before starting gRPC server

Blocked

- [x] Leader election #1204

### PR DESCRIPTION
**Pull request by stevesloka**
Updates #1280 by waiting for all informer caches must sync fully before starting Contour's gRPC server.

Big shoutout to @jpeach for help with the design!

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

File: cmd/contour/serve.go [hunk 1]
```diff
@@ -361,10 +389,8 @@ type informer interface {
 
 func startInformer(inf informer, log logrus.FieldLogger) func(stop <-chan struct{}) error {
 	return func(stop <-chan struct{}) error {
-		log.Println("waiting for cache sync")
-		inf.WaitForCacheSync(stop)
 
-		log.Println("started")
+		log.Println("starting informer")
```
Comment (davecheney) [inline cmd/contour/serve.go]: suggestion: maybe the next line should say "stopped informer", but that might be redundant as the context is already `context=coreinformers`

File: cmd/contour/serve.go [hunk 2]
```diff
@@ -303,9 +315,25 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 		log := log.WithField("context", "grpc")
+
+		log.Printf("waiting for informer caches")
+		cache.WaitForCacheSync(stop, synced...)
+		log.Printf("informer caches synced")
+
+		// Wait for cache to process
+		timeout := 15 * time.Second // Default timeout if reached code still progresses
+		if !eh.WaitForCacheProcess(timeout) {
+			log.Printf("!! timed out waiting %v for cache to process !!", timeout)
```
Comment (davecheney) [inline cmd/contour/serve.go]: what is the goal of this line? is this a warning? should contour wait longer if the timeout is reached?

File: internal/contour/handler.go [hunk 3]
```diff
@@ -267,3 +269,28 @@ func (e *EventHandler) setStatus(statuses map[dag.Meta]dag.Status) {
...
 }
+
+// WaitForCacheProcess waits for the Cache to be processed by checking the
+// value of "outstanding"
+// Return true if processed fully or false if timed out waiting
+func (e *EventHandler) WaitForCacheProcess(timeout time.Duration) bool {
+	ch := make(chan bool, 1)
+
+	go func() {
+		for e.outstanding > 0 {
```
Comment (davecheney) [inline internal/contour/handler.go]: I think this is a data race

File: internal/contour/handler.go [hunk 4]
```diff
@@ -267,3 +269,28 @@ func (e *EventHandler) setStatus(statuses map[dag.Meta]dag.Status) {
...
+			time.Sleep(100 * time.Millisecond) // ZzzzZZzzzz
+		}
+		ch <- true
```
Comment (davecheney) [inline internal/contour/handler.go]: you don't need to send on this channel, closing it is enough to make it readable by line 291

File: cmd/contour/serve.go [hunk 5]
```diff
@@ -303,9 +315,25 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
+	for _, inf := range informers {
+		synced = append(synced, inf.HasSynced)
+	}
+
 	// step 13. create grpc handler and register with workgroup.
 	g.Add(func(stop <-chan struct{}) error {
 		log := log.WithField("context", "grpc")
+
+		log.Printf("waiting for informer caches")
+		cache.WaitForCacheSync(stop, synced...)
```
Comment (jpeach) [inline cmd/contour/serve.go]: If `WaitForCacheSync` returns false, it means the stop chan was signalled and we should (log and) bail

https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L211

File: cmd/contour/serve.go (hunk 2 above)
Comment (stevesloka) [reply cmd/contour/serve.go]: I didn't want the wait for the processing to hold forever and block contour. My thought was to have a reasonable high default (this case 15 secs). If that time is reached then move along and start the gRPC server.

File: internal/contour/handler.go (hunk 3 above)
Comment (stevesloka) [reply internal/contour/handler.go]: Any thoughts on how we can check for when the cache queue is done? I have a cluster with 8k services, so those all load up and cause the dag to rebuild a bunch (1600+ times). 

This all doesn't take long, so maybe this check isn't needed but wanted to make sure that the dag that got served up to any Envoy had all the objects already processed.

File: cmd/contour/serve.go (hunk 2 above)
Comment (davecheney) [reply cmd/contour/serve.go]: But then the server starts and the cache isn't synced. Ideally we don't want to use a timeout, we want to issue a single operation then wait for it to complete.

File: internal/contour/handler.go (hunk 3 above)
Comment (davecheney) [reply internal/contour/handler.go]: I'm not sure polling outstanding is the right solution, even if we could solve the data race, as in a cluster with a stream of updates e.outstanding may never fall to zero, or it may fall to zero temporarily then rise.

Maybe its worth stepping back and asking the question about the operation's we want to occur before accepting gRPC requests. To my mind they are

1. Receive an initial list of updates from the api server, this is whatever client-go says it is by the "wait for cached" operation on the informer.
2. Initial at least one DAG rebuild cycle
3. Wait for that DAG cycle to be propagated through to the grpc caches.

For 1. I think this is well in hand with you change.
For 2. we have the "eh.UpdateNow" method
For 3. perhaps we need something similar to the eh.Sequence counter that is only tickled when cache updates happen. (eh.Sequence is tickled every time OnAdd/Update/Delete is called, which may not directly relate to cache updates)

File: cmd/contour/serve.go (hunk 2 above)
Comment (jpeach) [inline cmd/contour/serve.go]: I think there are 3 basic cases to handle:
1. The cache has not been built yet
2. The cache is still being rebuilt
3. The cache has been built (at least once)

`WaitForCacheProcess` is addressing (2), though the ordering of operations in the event handler means that it can succeed before the DAG has actually been sent to the cache. IIUC, we want to guarantee that we always fall into case (3). The event handler Sequence also doesn't have the semantics we want for this.

One option to consider is an event handler method that explicitly forces a cache rebuild. You'd have to call `UpdateNow()` and then need to have a way to block until the update is done. If the sequence counter was atomic, you could sleep wait until the sequence changed. It feels a bit hack but it would work :)

File: cmd/contour/serve.go (hunk 5 above)
Comment (stevesloka) [reply cmd/contour/serve.go]: 👍

**Comment by stevesloka** [comment]
This PR only now looks to wait for the k8s cache to sync. The other bits about waiting for the cachehandler to be processed will be addressed in a future PR.

File: cmd/contour/serve.go [hunk 6]
```diff
@@ -172,35 +175,45 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	// step 4. register our resource event handler with the k8s informers.
 	coreInformers.Core().V1().Services().Informer().AddEventHandler(eh)
```
Comment (davecheney) [inline cmd/contour/serve.go]: I think this is a duplicate of line 179

File: cmd/contour/serve.go [hunk 7]
```diff
@@ -172,35 +175,45 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 
 	// step 5. endpoints updates are handled directly by the EndpointsTranslator
 	// due to their high update rate and their orthogonal nature.
 	et := &contour.EndpointsTranslator{
 		FieldLogger: log.WithField("context", "endpointstranslator"),
 	}
-	coreInformers.Core().V1().Endpoints().Informer().AddEventHandler(et)
+
+	informers = append(informers, coreInformers.Core().V1().Endpoints().Informer())
+	informers[len(informers)-1].AddEventHandler(et)
```
Comment (davecheney) [inline cmd/contour/serve.go]: could this be clearer if it was written like this?
```
endpointInformer := coreInformers.Core().V1().Endpoints().Informer()
endpointInformer.AddEventhandler(et)
informers = append(informers, endpointInformer
```

File: cmd/contour/serve.go [hunk 8]
```diff
@@ -172,35 +175,45 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 
 	// Add informers for each root-ingressroute namespaces
 	for _, inf := range namespacedInformers {
-		inf.Core().V1().Secrets().Informer().AddEventHandler(eh)
+		informers = append(informers, inf.Core().V1().Secrets().Informer())
 	}
 	// If root-ingressroutes are not defined, then add the informer for all namespaces
 	if len(namespacedInformers) == 0 {
-		coreInformers.Core().V1().Secrets().Informer().AddEventHandler(eh)
+		informers = append(informers, coreInformers.Core().V1().Secrets().Informer())
```
Comment (davecheney) [inline cmd/contour/serve.go]: Maybe a small helper that took an informer and a resource handler and returned the informer would help
```
informers = append(informers, registerEventHandler(inf.Core().V1().Secrets().Informer())
...
func registerEventHandler(inf <informer type>, eh cache.ResourceEventHandler) <informer type> {
      inf.AddEventHandler(eh)
      return inf
}
```
Then every time we add an informer to the slice, we can also register an event handler at the same time

File: internal/contour/handler.go [hunk 9]
```diff
@@ -62,6 +62,10 @@ type EventHandler struct {
 	// will be made once CacheHandler.OnUpdate has been called.
 	Sequence chan int
 
+	// outstanding counts the number of events received but not
```
Comment (davecheney) [inline internal/contour/handler.go]: Please revert the changes to this file

File: cmd/contour/serve.go [hunk 10]
```diff
@@ -303,9 +316,21 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
+	synced := make([]cache.InformerSynced, 0, len(informers))
+	for _, inf := range informers {
+		synced = append(synced, inf.HasSynced)
+	}
+
 	// step 13. create grpc handler and register with workgroup.
 	g.Add(func(stop <-chan struct{}) error {
 		log := log.WithField("context", "grpc")
+
+		log.Printf("waiting for informer caches")
```
Comment (davecheney) [inline cmd/contour/serve.go]: waiting for informer caches to sync

File: cmd/contour/serve.go [hunk 11]
```diff
@@ -303,9 +316,21 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	eh.Metrics = metrics
 	eh.CacheHandler.Metrics = metrics
 
+	synced := make([]cache.InformerSynced, 0, len(informers))
+	for _, inf := range informers {
```
Comment (davecheney) [inline cmd/contour/serve.go]: Possibly move these four lines closer to where they are used inside the step 13 function below.

**Comment by stevesloka** [comment]
Comments addressed @davecheney, thanks!

File: cmd/contour/serve.go [hunk 12]
```diff
@@ -170,37 +173,40 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	}
 
 	// step 4. register our resource event handler with the k8s informers.
-	coreInformers.Core().V1().Services().Informer().AddEventHandler(eh)
+	informers := []cache.SharedIndexInformer{
```
Comment (davecheney) [inline cmd/contour/serve.go]: should these be passed through `registerEventHandler`?

Comment (stevesloka) [reply cmd/contour/serve.go]: I have `informers` getting passed through now, is this in line with what you were suggesting?

File: cmd/contour/serve.go [hunk 13]
```diff
@@ -306,6 +311,18 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
...
 		log := log.WithField("context", "grpc")
+
+		synced := make([]cache.InformerSynced, 0, len(informers))
+		for _, inf := range informers {
+			synced = append(synced, inf.HasSynced)
+		}
+
+		log.Printf("waiting for informer caches to sync")
+		if !cache.WaitForCacheSync(stop, synced...) {
+			return fmt.Errorf("error waiting for cache to sync")
```
Comment (jpeach) [inline cmd/contour/serve.go]: Nit: IIUC this isn't really an error, it's that the stop chan was signalled.

File: cmd/contour/serve.go (hunk 12 above)
Comment (davecheney) [reply cmd/contour/serve.go]: I wasn't able to find the place where those 6 informers added to `informers` had an event handler attached to them, but I think you resolved this in d8fb549
